---
title: "Project 3"
output: html_notebook
---

# Quiz 3

## Question 1: 

What is the mean expression across all features for sample 5 in the ALL dataset (from the ALL package)?

```{r}

ALL # load the data in
# sample 5 matrix
mean(exprs(ALL[,5]))
```

## Question 2: 

We will use the biomaRt package to annotate an Affymetrix microarray. We want our results in the hg19 build of the human genome and we therefore need to connect to Ensembl 75 which is the latest release on this genome version. How to connect to older versions of Ensembl is described in the biomaRt package vignette; it can be achived with the command mart <- useMart(host='feb2014.archive.ensembl.org', biomart = "ENSEMBL_MART_ENSEMBL")mart \<- useMart(host=’feb2014.archive.ensembl.org’, biomart = "ENSEMBL_MART_ENSEMBL").

**Question:** Using this version of Ensembl, annotate each feature of the ALL dataset with the Ensembl gene id. How many probesets (features) are annotated with more than one Ensembl gene id?

```{r}

# get the mart 
mart <- useMart(host='feb2014.archive.ensembl.org', biomart = "ENSEMBL_MART_ENSEMBL")
# connect to ensemble and get the dataset 
ensembl <- useDataset("hsapiens_gene_ensembl", mart)
features <- featureNames(ALL) # get all the names of the features (non ensembled  here)

listAttributes(ensembl) # ensemble provided gene_id ensembl_gene_id

# get the annotation: the ensembl gene ids 
annotated_ALL <- getBM(
  attributes = c("ensembl_gene_id", "affy_hg_u95av2"), # output columsn 
  filters = "affy_hg_u95av2", # query the id from above
  values = features, # annotate with the 
  mart = ensembl # the mart used 
)

annotated_ALL #view the annotated dataset 

# figure out how many of the features occur twice? 
sum( table(annotated_ALL$affy_hg_u95av2) > 1 ) 
```

## Question 3: 

How many probesets (Affymetrix IDs) are annotated with one or more genes on the autosomes (chromosomes 1 to 22).

```{r}

# just have to add further annotation: another colum with chromosome so that we only have stuff on the autosomes
# mirror what we did in lecture

chromosomes <- c(1:22) # have the chromosomes, only autosomes

listAttributes(ensembl) # chromsome_name is the available metadata in the mart

annotated_2_ALL <- getBM(
  attributes = c("ensembl_gene_id", "affy_hg_u95av2", "chromosome_name"), # output columsn 
  filters = c("affy_hg_u95av2", "chromosome_name"), # query the id from above but with 2 args this time 
  values = list(features, chromosomes), # annotate with both of these 
  mart = ensembl # the mart used is the same one 
)

annotated_2_ALL

# we just need all the records in this dataset 
table(annotated_2_ALL[,2]) # every feature showed up x times
table(table(annotated_2_ALL[,2])) # the showing up x times showed up y times, given in this table 
sum (table(table(annotated_2_ALL[,2])))
```

##  Question 4

Use the MsetEx dataset from the minfiData package. Part of this question is to use the help system to figure out how to address the question.

**Question:** What is the mean value of the Methylation channel across the features for sample “5723646052_R04C01”?

```{r}
# installed the minfiData package and minfi
library(minfiData)
library(minfi) # came with the minfiData package 

class(MsetEx) # look for the package in minfi, it's a "MethylSet" 
# looked up what kinds of operations on a MSet

methylation_matrix <- getMeth(MsetEx) # get the methylation data matrix
featureNames(MsetEx) # these are the features
sampleNames(MsetEx) # the stuff being measured (encoded in soem way)
# only want 5723646052_R04C01
mean(methylation_matrix[,2]) # col 2 mean value
```

## Question 5

**Question:** Access the processed data from NCBI GEO Accession number GSE788. What is the mean expression level of sample GSM9024?

```{r}

library(GEOquery)

# get datasets into list from accession number 
eList = getGEO("GSE788")
eList # GSM9024 listed in samples 

GSE788 <- eList[[1]]
GSE788 # expression set 

exprs(GSE788) # expression matrix 
GSM9024_mean <- mean (exprs(GSE788)[,2]) # second sample 
GSM9024_mean
```

## Question 6

We are using the airway dataset from the airway package.

**Question:** What is the average of the average length across the samples in the expriment?

```{r}

library(airway)
airway # summarized expert 

# everything here has a granges, we just need to accesss it 
rownames(airway)
colnames(airway)

colData(airway) # avgLength is one of the columns 

mean ( airway$avgLength )
```

## Question 7: 

We are using the airway dataset from the airway package. The features in this dataset are Ensembl genes.

**Question:** What is the number of Ensembl genes which have a count of 1 read or more in sample SRR1039512?

```{r}

# the assays part has the counts in it 
SRR1039512 <- assay(airway)[,3]
class(SRR1039512 )

sum(SRR1039512 >= 1) # everyone with 1 or more read 
```

## Question 8

**Question:** The airway dataset contains more than 64k features. How many of these features overlaps with transcripts on the autosomes (chromosomes 1-22) as represented by the TxDb.Hsapiens.UCSC.hg19.knownGene package?

**Clarification:** A feature has to overlap the actual transcript, not the intron of a transcript. So you will need to make sure that the transcript representation does not contain introns.

```{r}

txdb # still in my R environment somewhere lmao 
txdb$user_seqlevels

# need trancripts on the autosomes 
chr1to22 <- paste0("chr", 1:22) # no spaces paste0
# only keep the transcripts that were on the autosomes
txdb_autosome <- keepSeqlevels(
  exons(txdb), # no introns, means only exons 
  chr1to22, pruning.mode = "coarse")
txdb_autosome

# rename the txdb autosome with ncbi style so there's something to merge with the airway
txdb_renamed <- mapSeqlevels( seqlevels(txdb_autosome), "NCBI")
txdb_exons_renamed <- renameSeqlevels (txdb_autosome, txdb_renamed)
txdb_exons_renamed

# get the actual overlaps between airway and exons 
subsetByOverlaps(airway, txdb_exons_renamed) # 26276 records
```

## Question 9

The expression measures of the airway dataset are the number of reads mapping to each feature. In the previous question we have established that many of these features do not overlap autosomal transcripts from the TxDb.Hsapiens.UCSC.hg19.knownGene. But how many reads map to features which overlaps these transcripts?

**Question:** For sample SRR1039508, how big a percentage (expressed as a number between 0 and 1) of the total reads in the airway dataset for that sample, are part of a feature which overlaps an autosomal TxDb.Hsapiens.UCSC.hg19.knownGene transcript?

```{r}

colnames(airway)

SRR1039508 <- airway[, 1]
# only overlaps between the exons in the autosome and the sample 
SRR1039508_autosome <- subsetByOverlaps(SRR1039508, txdb_exons_renamed) 

autosome_reads <- sum ( assay(SRR1039508_autosome, "counts") ) # counts num reads 
reads <- sum ( assay(SRR1039508, "counts") )

autosome_reads / reads
```

## Question 10 

Consider sample SRR1039508 and only consider features which overlap autosomal transcripts from TxDb.Hsapiens.UCSC.hg19.knownGene. We should be able to very roughly divide these transcripts into expressed and non expressed transcript. Expressed transcripts should be marked by H3K4me3 at their promoter. The airway dataset have assayed “airway smooth muscle cells”. In the Roadmap Epigenomics data set, the E096 is supposed to be “lung”. Obtain the H3K4me3 narrowPeaks from the E096 sample using the AnnotationHub package.

**Question:** What is the median number of counts per feature (for sample SRR1039508) containing a H3K4me narrowPeak in their promoter (only features which overlap autosomal transcripts from TxDb.Hsapiens.UCSC.hg19.knownGene are considered)?

**Clarification:** We are using the standard 2.2kb default Bioconductor promotor setting.

**Conclusion** Compare this to the median number of counts for features without a H3K4me3 peak. Note that this short analysis has not taken transcript lengths into account and it compares different genomic regions to each other; this is highly suscepticle to bias such as sequence bias.

```{r}

# tried couple different approaches, couldn't get what the question was asking and didn't end up with one of the multiple choice options... 
```
