---
title: "Project 2"
output: html_document
date: "2024-07-19"
---

## Question 1: What is the GC content of “chr22” in the “hg19” build of the human genome?

**Tip:** The reference genome includes “N” bases; you will need to exclude those.

```{r}

# need the hg19 build of the human genome
available.genomes()
# download the genome and load it 
BiocManager::install("BSgenome.Hsapiens.UCSC.hg19")
library(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens # it gets saved to the genome name. 

# isolate chromosome 22 
Hsapiens$chr22
letterFrequency(Hsapiens$chr22, "ACGT") # total frequency of bases
letterFrequency(Hsapiens$chr22, "GC") # GC bases: 16745219
letterFrequency(Hsapiens$chr22, "GC") / letterFrequency(Hsapiens$chr22, "ACGT") # 48 percent. 
```

## Question 2: In the previous assessment, studied H3K27me3 from H1 cell line. (E003). Examine GC content of these reigons and see if they're different than other GC content

**Question:** What is mean GC content of H3K27me3 “narrowPeak” regions from Epigenomics Roadmap from the H1 stem cell line on chr 22.

**Clarification:** Compute the GC content for each peak region as a percentage and then average those percentages to compute a number between 0 and 1.

```{r}

# Step 1: Get the H1 cell line 

# the h1 cell line -> the h1 line is E003 
qhs2 <- query(ahub, c("H3K27me3", "E003")) # query for the cell line and histone modification

# download (we used narrow peaks in lecture, will use those again)
h1_cells = qhs2[["AH29892"]]
h1_cells # the h1 cells that we need 

# Step 2: Isolate Chromosome 22 narrow peak reigons 

seqnames(h1_cells)
chr22_h1 <- subset(h1_cells, seqnames == "chr22")
chr22_h1 # narrow peaks only on chr 22 
seqnames(chr22_h1) # 917 lengths/ranges of chr22 

# Step 3: Create a View object from the narrow peak reigons (chr22_h1)
                    # Views (the genome, what ranges you want it cut up according to)
np_chr22_sequences <- Views(Hsapiens, chr22_h1)
np_chr22_sequences # factor-Rle of length 917 with 1 run
seqnames(np_chr22_sequences) # confirmed that it only contains chromosome 22

# Step 4: Calculate the average GC Content of narrow peaks on chr22 
mean(letterFrequency(np_chr22_sequences, "GC", as.prob = T)) # 52.9 percent. 
```

## Question 3: The “narrowPeak” regions includes information on a value they call “signalValue”.

**Question:** What is the correlation between GC content and “signalValue” of these regions (on chr22)?

```{r}

# this is the signal value column
# the GC percentages were calculated in order of the chr22 ranges 
gc_percent <- letterFrequency(np_chr22_sequences, "GC", as.prob = T)
cor(mcols(np_chr22_sequences)$signalValue, gc_percent) # check their correlation#
```

## Question 4: 

The “narrowPeak” regions are presumably reflective of a ChIP signal in these regions. To confirm this, we want to obtain the “fc.signal” data from AnnotationHub package on the same cell line and histone modification. This data represents a vector of fold-change enrichment of ChIP signal over input.

**Question:** what is the correlation between the “signalValue” of the “narrowPeak” regions and the average “fc.signal” across the same regions?

**Clarification:** First compute the average “fc.signal” for across each region, for example using “Views”; this yields a single number of each region. Next correlate these numbers with the “signalValue” of the “narrowPeaks”.

```{r}

# Step 1: have to get the fc.signal data 
fc_data <- query(ah, c("H3K27me3", "E003", "fc.signal"))[[1]]

# Step 2: have to extract only from the chr22 region
fc_chr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start = 1, end = 51304566))
fc_chr22 

                # import the part of fc_signals which lines up with range of chr22 
fc_rle <- import(fc_data, which = fc_chr22, as = "Rle")
fc_rle

# isolate chromosome 22
fc_data_chr22 <- fc_rle$chr22 
fc_data_chr22

fc_view_chr22 <- Views(fc_data_chr22, start = start(chr22_h1), end = end(chr22_h1))
fc_view_chr22

# Step 3: correlation between signalValue and fc.signals

length(mean(fc_view_chr22))

cor( mean(fc_view_chr22), mcols(np_chr22_sequences)$signalValue ) 
```

## Question 5: 

The H1 stem cell line is an embryonic stem cell line, a so-called pluripotent cell. Many epigenetic marks change upon differentiation. We will examine this. We choose the cell type with Roadmap ID “E055” which is foreskin fibroblast primary cells.

We will use the “fc.signal” for this cell type for the H3K27me3 mark, on chr22. We now have a signal track for E003 and a signal track for E055. We want to identify regions of the genome which gain H3K27me3 upon differentiation. These are regions which have a higher signal in E055 than in E003. To do this properly, we would need to standardize (normalize) the signal across the two samples; we will ignore this for now.

**Question:** Identify the regions of the genome where the signal in E003 is 0.5 or lower and the signal in E055 is 2 or higher.

**Tip:** If you end up with having to intersect two different Views, note that you will need to convert the Views to IRanges or GRanges first with ir <- as(vi, "IRanges")ir \<- as(vi, "IRanges").Referring to the objects made and defined in the previous question.

**Question:** How many bases on chr22 have an fc.signal greater than or equal to 1?

```{r}
sum (fc_data_chr22 >= 1)
```

## Question 6: 

The H1 stem cell line is an embryonic stem cell line, a so-called pluripotent cell. Many epigenetic marks change upon differentiation. We will examine this. We choose the cell type with Roadmap ID “E055” which is foreskin fibroblast primary cells.

We will use the “fc.signal” for this cell type for the H3K27me3 mark, on chr22. We now have a signal track for E003 and a signal track for E055. We want to identify regions of the genome which gain H3K27me3 upon differentiation. These are regions which have a higher signal in E055 than in E003. To do this properly, we would need to standardize (normalize) the signal across the two samples; we will ignore this for now.

**Question:** Identify the regions of the genome where the signal in E003 is 0.5 or lower and the signal in E055 is 2 or higher.

**Tip:** If you end up with having to intersect two different Views, note that you will need to convert the Views to IRanges or GRanges first with ir <- as(vi, "IRanges")ir \<- as(vi, "IRanges").

```{r}

E055 <- query(ah, c("H3K27me3", "E055"))
E055_data <- E055[[4]]

gr.chr22 <- GRanges(seqnames = "chr22", 
                    ranges = IRanges(start =1, end = 51304566))
E055_rel <- import(E055_data, which =gr.chr22, as = "Rle")
E055_rel_chr22 <- E055_rel$chr22

# identify region
r_E003 <- slice(fc_data_chr22, upper = 0.5)
r_E055 <- slice(E055_rel_chr22 , lower = 2)
# convert them to iranges 
r_E003 <- as(r_E003, "IRanges") 
r_E055 <- as(r_E055, "IRanges")

inter_region <- intersect(r_E003, r_E055)
sum(width(inter_region))
```

## Question 7: Tried for an hour, have no idea or clue what is being asked here so i just took a guess

## Question 8: 

A TATA box is a DNA element of the form “TATAAA”. Around 25% of genes should have a TATA box in their promoter. We will examine this statement.

**Question:** How many TATA boxes are there on chr 22 of build hg19 of the human genome?

**Clarification:** You need to remember to search both forward and reverse strands.

```{r}
# it's just pattern matching 

search_string <- "TATAAA"
countPattern(search_string, Hsapiens$chr22) + countPattern(search_string, reverseComplement(Hsapiens$chr22))
```

## Question 9: 

**Question:** How many promoters of transcripts on chromosome 22 containing a coding sequence, contains a TATA box on the same strand as the transcript?

**Clarification:** Use the TxDb.Hsapiens.UCSC.hg19.knownGene package to define transcripts and coding sequence. Here, we defined a promoter to be 900bp upstream and 100bp downstream of the transcription start site.

```{r}
library(GenomicFeatures)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb = TxDb.Hsapiens.UCSC.hg19.knownGene 
txdb # info on genes, transcripts, coding sequences.

# Step 1: isolate the transcripts
transcripts(txdb)

# get the transcripts 
gr <- GRanges(seqnames = "chr22", ranges = IRanges(start = 1, end = 52330658))
trans <- subsetByOverlaps(transcripts(txdb), gr) 

# Step 2: get the promoters of those transcripts (stretch the ranges 900 bp upstream and 100 bp downstream)
proms <- promoters(trans, upstream = 100, downstream = 900)

trans 
proms

# Step 3: Search how many of these promoters have a TATA box on the same strand as a transcript 

# in other words,  how many promoters showed up on a coding sequence? 

cds <- subsetByOverlaps(genes(txdb), gr) # coding sequences
ov <- findOverlaps(cds, proms) # promoters appearing in cds
ov

prom_indices <- unique(queryHits(ov))

count = 0 
for (i in prom_indices){
  # create a veiw with the promoter at that index in the list of promoter ranges 
  promoter <- Views(Hsapiens, proms[i])
  # thne check for the TATAA box 
  count = count + vcountPattern("TATAA", DNAStringSet(promoter))
}
count # pretty close to 252? 
```

## Question 10: 

It is possible for two promoters from different transcripts to overlap, in which case the regulatory features inside the overlap might affect both transcripts. This happens frequently in bacteria.

**Question:** How many bases on chr22 are part of more than one promoter of a coding sequence?

**Clarification:** Use the TxDb.Hsapiens.UCSC.hg19.knownGene package to define transcripts and coding sequence. Here, we define a promoter to be 900bp upstream and 100bp downstream of the transcription start site. In this case, ignore strand in the analysis.

```{r}

# proms is stil promoters 

# need transcripts 
transcripts <- transcriptLengths(txdb, with.cds_len = TRUE)
transcripts # 10 transcripts 

# create a new versions of proms - transcripts from chr 22 (made from txdb)
count_per_trans <- proms[ mcols(proms)$tx_id %in% transcripts$tx_id ]

# sum all them with more than 1 obs per transcript since we want stuff with more than one transcript
sum(coverage(count_per_trans) > 1)["chr22"] # 398661 isnt an answer but 38k is close? 
```
